File: programmers/PCCP 기출문제/BandageWrapping.java
// 1번 붕대 감기
// 2025-09-01 Mon
class BandageWrapping {
    public int solution(int[] bandage, int health, int[][] attacks) {
        int castTime = bandage[0];
        int healPerSecond = bandage[1];
        int extraHeal = bandage[2];
        int maxHealth = health;
        int lastAttackTime = 0;

        int timeGap;
        int extraHealCount;
        for (int[] attack : attacks) {
            timeGap = attack[0] - lastAttackTime - 1;
            extraHealCount = timeGap / castTime;
            lastAttackTime = attack[0];
            
            health = Math.min(maxHealth, health + (timeGap * healPerSecond));
            health = Math.min(maxHealth, health + (extraHealCount * extraHeal));

            health -= attack[1];

            if (health <= 0) {
                return -1;
            }
        }

        return health;
    }

    public static void main(String[] args) {
        BandageWrapping solution = new BandageWrapping();

        int answer = solution.solution(new int[]{5, 1, 5}, 30, new int[][]{{2, 10}, {9, 15}, {10, 5}, {11, 5}});

        System.out.println(answer);
    }
}

File: programmers/PCCP 기출문제/review/BandageWrapping.py
def solution(bandage, health, attacks):
    cast_time = bandage[0]
    heal_per_second = bandage[1]
    extra_heal = bandage[2]
    max_health = health
    last_attack_time = 0

    for attack_time, damage in attacks:
        time_gap = attack_time - last_attack_time - 1
        extra_heal_count = time_gap // cast_time
        last_attack_time = attack_time

        health = min(max_health, health + (time_gap * heal_per_second))
        health = min(max_health, health + (extra_heal_count * extra_heal))

        health -= damage

        if health <= 0:
            return -1

    return health
